### ISSUE TITLE
How do we use sceneform along with ARCloud anchors?

### ISSUE THREAD

**Issue by rohitagarwal3011** [google-ar/arcore-android-sdk#442]
Every time I try to host an anchor "**session.hostCloudAnchor(anchor);**" it shows up **NotTrackingException**. How can we host and resolve the Anchor that we get from  **arFragment.setOnTapArPlaneListener** .

**Comment by rohitagarwal3011** [comment]
This is the code i am using,

         arFragment.setOnTapArPlaneListener(
        (HitResult hitResult, Plane plane, MotionEvent motionEvent) -> {
          Camera camera = arFragment.getArSceneView().getArFrame().getCamera();
          TrackingState cameraTrackingState = camera.getTrackingState();

          if (andyRenderable == null) {
            return;
          }

          if (plane.getType() != Type.HORIZONTAL_UPWARD_FACING) {
            return;
          }

          if (cameraTrackingState == TrackingState.TRACKING && session!= null) {
            // Create the Anchor.
            anchor = hitResult.createAnchor();
            try{
            session.hostCloudAnchor(anchor);
            }
            catch (NotTrackingException e)
            {
                e.printStackTrace();

            }
            setNewAnchor(anchor);
            appAnchorState = AppAnchorState.HOSTING;
            Toast.makeText(HelloSceneformActivity.this, "Now, hosting anchor", Toast.LENGTH_SHORT)
                .show();

            AnchorNode anchorNode = new AnchorNode(anchor);
            anchorNode.setParent(arFragment.getArSceneView().getScene());

            // Create the transformable andy and add it to the anchor.
            TransformableNode andy = new TransformableNode(arFragment.getTransformationSystem());
            andy.setParent(anchorNode);
            andy.setRenderable(andyRenderable);
            andy.select();

            checkUpdatedAnchor();
          }
        });``

**Comment by inio** [comment]
Where are you getting `session`?  Are you creating it separately or calling `arFragment.getArSceneView().getSession()` (or I guess creating it and then using `ArSceneView.setupSession()`)?

**Comment by rohitagarwal3011** [comment]
I am creating it seperately.

    @Override
    protected void onResume() {
        super.onResume();
        if(session==null)
        {
            //Toast.makeText(HelloSceneformActivity.this,"Session is null",Toast.LENGTH_SHORT).show();
            try {
                session = new Session(/* context= */ this);
            } catch (UnavailableArcoreNotInstalledException e) {
                e.printStackTrace();
            } catch (UnavailableApkTooOldException e) {
                e.printStackTrace();
            } catch (UnavailableSdkTooOldException e) {
                e.printStackTrace();
            }

            // Create default config and check if supported.
            Config config = new Config(session);
            if (!session.isSupported(config)) {
//                showSnackbarMessage("This device does not support AR", true);
            }
            config.setCloudAnchorMode(Config.CloudAnchorMode.ENABLED);
            session.configure(config);
        }
        else {
           // Toast.makeText(HelloSceneformActivity.this,"Session is not null",Toast.LENGTH_SHORT).show();
        }

        try {
            session.resume();
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

**Comment by inio** [comment]
I see you're not setting `LATEST_CAMERA_IMAGE`, which makes me wonder if you're using [`ArSceneView.setupSession()`](https://developers.google.com/ar/reference/java/com/google/ar/sceneform/ArSceneView#setupSession\(com.google.ar.core.Session\)).  I also see a catch-and-forget on `session.resume()`, which makes me think that method is failing.

You can only have one running Session at a time.  If you let Sceneform try to start one yours definitely won't start.  You need to give your configured session to sceneform using `ArSceneView.setupSession()`.

**Comment by claywilkinson** [comment]
It looks like there is some code missing so I am not sure what's the problem.  I was able to successfully convert the sample to use Sceneform in my [personal fork](https://github.com/claywilkinson/arcore-android-sdk/pull/1).   Feel free to compare your code to it.

I'll work on getting the converted sample added to the Sceneform SDK....

**Comment by claywilkinson** [comment]
Interesting, I'll take a look and see if I can reproduce it.  The fix is to check for null before detaching, if it is null, there is no need to detach :)
```
 if (anchorNode != null) {
      if(anchorNode.getAnchor() != null) {
         anchorNode.getAnchor().detach();
      }
      for (Node node : anchorNode.getChildren()) {
        node.setParent(newAnchorNode);
      }
      anchorNode.setParent(null);
```

As for the resolveCloudAnchor() and the session being null, I'll have to look more closely, but it sounds like the session is not setup yet for some reason...

**Comment by rohitagarwal3011** [comment]
Okay.Thanks for the correction. 
Is there are a way that we can attach multiple objects to a single anchor?

**Comment by claywilkinson** [comment]
Yes, you can create any number of Nodes and set the parent to the anchor node.  The Sceneform scene is a tree, the root is the Scene object, and can have multiple child nodes.

**Comment by rohitagarwal3011** [comment]
Please correct me if I am wrong, 
Scene object is the parent to an AnchorNode and we can add multiple nodes to an AnchorNode.

**Comment by rohitagarwal3011** [comment]
Okay. So you mean to say if store the Pose of all the child nodes and then retrieve them then all the objects attached to that AnchorNode can be resolved?
Does ARCore provide the necessary information for each node that they can be stored and restored across devices?

**Comment by claywilkinson** [comment]
you can  get the rotation and position of all the nodes.  It is a little more involved that can be easily answered here.  You could look at the [Just A Line](https://github.com/googlecreativelab/justaline-android)  demo app, specifically how the [FirebaseAnchorData](https://github.com/googlecreativelab/justaline-android/blob/master/app/src/main/java/com/arexperiments/justaline/model/FirebaseAnchorData.java) class is used.

**Comment by rohitagarwal3011** [comment]
Okay thanks. I will definitely go through it and get back to you in case of any help.
Just one last doubt,
Is there a way in which we can store the whole information of AnchorNode and other nodes in our own database instead of CloudAnchor especially because of 24 hours limitation?

**Comment by claywilkinson** [comment]
At this time there is no way to privately store cloud anchors.
